src/mptcpize.c

```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: BSD-3-Clause
+#define  _GNU_SOURCE
+
+#include <linux/limits.h>
+
+#include <dlfcn.h>
+#include <errno.h>
+#include <error.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+void help(void)
+{
+	fprintf(stderr, "mptcpize <cmd>\n"
+		"\tavaliable <cmd>:\n"
+		"\trun [-d] prog [<args>]    run target program with specified arguments\n"
+		"\t\t\t forcing MPTCP socket usage instead of TCP\n"
+		"\t\t\t if the '-d' argument is provided, dump messages on stderr\n"
+		"\t\t\t when a TCP socket is forced to MPTCP\n"
+		"\tservice <unit>       create/update <unit>_mptcp systemd unit, which\n"
+		"\t\t\t force the given service to run under the above launcher\n"
+		"\thelp                 show this help\n");
+}
+
+#define MPTCPWRAP_ENV "LD_PRELOAD="PKGLIBDIR"/libmptcpwrap.so.0.0."LIBREVISION
+
+static int run(int argc, char *av[])
+{
+	int i, nr = 0, debug = 0;
+	char **envp, **argv;
+
+	if (argc > 0 && !strcmp(av[0], "-d")) {
+		debug = 1;
+		argc--;
+		av++;
+	}
+
+	if (argc < 1) {
+		fprintf(stderr, "missing command argument");
+		help();
+		return -1;
+	}
+
+	// build environment, copying the current one ...
+	while (environ[nr])
+		nr++;
+	envp = calloc(nr + 3, sizeof(char *));
+	if (!envp)
+		error(1, errno, "can't allocate env list");
+
+	// ... filtering out any 'LD_PRELOAD' ...
+	nr = 0;
+	i = 0;
+	while (environ[nr]) {
+		if (strncmp(environ[nr], "LD_PRELOAD=", 11)) {
+			envp[i] = environ[nr];
+			i++;
+		}
+		nr++;
+	}
+
+	// ... appending the mptcpwrap preload...
+	envp[i++] = MPTCPWRAP_ENV;
+
+	// ... and enable dbg if needed
+	if (debug)
+		envp[i++] = "MPTCPWRAP_DEBUG=1";
+
+	// build the NULL terminated arg list
+	argv = calloc(argc + 1, sizeof(char *));
+	if (!argv)
+		error(1, errno, "can't allocate argument list");
+
+	memcpy(argv, av, argc * sizeof(char*));
+	return execvpe(argv[0], argv, envp);
+}
+
+#define SYSTEMCTL_SHOW "systemctl show -p FragmentPath "
+
+static char *locate_unit(const char *name)
+{
+	char *cmd, *line = NULL;
+	FILE *systemctl;
+	size_t len = 0;
+	ssize_t read;
+
+	if (!access(name, R_OK))
+		return strdup(name);
+
+	len = strlen(name) + 1 + strlen(SYSTEMCTL_SHOW);
+	cmd = malloc(len);
+	if (!cmd)
+		error(1, 0, "can't allocate systemctl command string");
+	sprintf(cmd, SYSTEMCTL_SHOW"%s", name);
+
+	systemctl = popen(cmd, "r");
+	if (!systemctl)
+		error(1, errno, "can't execute %s", cmd);
+
+	while ((read = getline(&line, &len, systemctl)) != -1) {
+		if (!strncmp(line, "FragmentPath=", 13)) {
+			char *ret = strdup(&line[13]);
+
+			// trim trailing newline, if any
+			len = strlen(ret);
+			if (len > 0 && ret[len - 1] == '\n')
+				ret[len - 1] = 0;
```

Comment:
should we check if we can write in this file? → `access(...) == 0`

Reply:
uhm ?!? here we are reading from a pipe?!? write operation will happen later in unit_update(). I think checks are in place there...

Reply:
yes, it is because we will ultimately modify the file. But yes, we will detect the error later, no need to change :)